cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1035,6 +1025,26 @@ class ObjectOutputStream final : public io::OutputStream {
     return Status::OK();
   }
 
+  static void HandleUploadOutcome(const std::shared_ptr<UploadState>& state,
+                                  int part_number, const S3Model::UploadPartRequest& req,
+                                  const Result<S3Model::UploadPartOutcome>& result) {
+    std::unique_lock<std::mutex> lock(state->mutex);
+    if (!result.ok()) {
+      state->status &= result.status();
+    } else {
+      const auto& outcome = *result;
+      if (!outcome.IsSuccess()) {
+        state->status &= UploadPartError(req, outcome);
+      } else {
+        AddCompletedPart(state, part_number, outcome.GetResult());
+      }
+    }
+    // Notify completion, regardless of success / error status
+    if (--state->parts_in_progress == 0) {
```

Comment:
Similar to the other comment, if you use `Flush()` it blocks.  You could create a `FlushAsync` and update the code here so that instead of just `notify_all()` it marked a future finished.  Or you could use a threaded task group which already has this logic.

Reply:
For now we're not exposing any `FlushAsync` method on `OutputStream`. I don't know if that would later be necessary. But we can indeed prepare for it by using a future.